apps/heft/src/cli/actions/BuildAction.ts

```diff
@@ -74,17 +74,69 @@ export interface ICopyStaticAssetsConfiguration extends ISharedCopyStaticAssetsC
   destinationFolderNames: string[];
 }
 
+/**
+ * @public
+ */
+export interface IEmitModuleKindBase<TModuleKind> {
+  moduleKind: TModuleKind;
+  outFolderPath: string;
+}
+
+/**
+ * @public
+ */
+export type IEmitModuleKind = IEmitModuleKindBase<
+  'commonjs' | 'amd' | 'umd' | 'system' | 'es2015' | 'esnext'
+>;
+
+/**
+ * @public
+ */
+export type CopyFromCacheMode = 'hardlink' | 'copy';
+
+/**
+ * @public
+ */
+export interface ISharedTypescriptConfiguration {
+  /**
+   * Can be set to 'copy' or 'hardlink'. If set to 'copy', copy files from cache. If set to 'hardlink', files will be
+   * hardlinked to the cache location. This option is useful when producing a tarball of build output as TAR files
+   * don't handle these hardlinks correctly. 'hardlink' is the default behavior.
+   */
+  copyFromCacheMode?: CopyFromCacheMode | undefined;
+
+  /**
+   * If provided, emit these module kinds in addition to the modules specified in the tsconfig.
+   * Note that this option only applies to the main tsconfig.json configuration.
+   */
+  additionalModuleKindsToEmit?: IEmitModuleKind[] | undefined;
+}
+
+/**
+ * @public
+ */
+export interface ITypescriptConfiguration extends ISharedTypescriptConfiguration {
+  tsconfigPaths: string[];
+  tslintConfigPath: string | undefined;
+  isLintingEnabled: boolean | undefined;
```

Comment:
How will TSLint vs ESLint be handled with this design?

Reply:
It'll really depend on if ESLint can take an existing TypeScript analysis or not. If it can, it should live parallel to TSLint. If it can't, there's no reason to marry it to the TypeScript builder.

There's also no technical reason a project can't use both TSLint and ESLint. I can't think of a reason anyone would want to do that, but it's totally technically possible.

Reply:
@iclanton  We actually have some projects at work that invoke ESLint and TSLint together, because the person who set it up didn't think it through, and it never got fixed. It's a confusing experience. Maybe it could be allowed to facilitate a migration, but it shouldn't be easy to do by mistake.

Reply:
> It'll really depend on if ESLint can take an existing TypeScript analysis or not.

This is an interesting design question.

**Pro:**
- I don't think ESLint supports this today, but the [typescript-eslint project](https://github.com/typescript-eslint/typescript-eslint/) is maintained by a different group that tend to be quite fast with turning around PRs (since it is still relatively new), so we could realistically contribute this feature.
- The `ts.Program` [gets created here](https://github.com/typescript-eslint/typescript-eslint/blob/7c7032322f55d9492e21d3bfa5da16da1f05cbce/packages/typescript-estree/src/create-program/createDefaultProgram.ts#L20), and the semantic analysis [happens like this](https://github.com/typescript-eslint/typescript-eslint/blob/3a15413d87a3429ebf19af2cc5db76c9e7ffe4e7/packages/typescript-estree/src/semantic-or-syntactic-errors.ts#L28), which would be pretty easy to hook into. Also, the scenario we care about optimizing invokes ESLint exactly once, without any watch or updates. So the work involved seems not expensive.

**Con:**
- On the other hand, in a Rush repo the big win for ESLint would be the multiphase feature, where linting would happen in the background as a later phase that does not block downstream projects. With this optimization, the actual CPU cost of linting might matter less.